### PR TITLE
BUG: combine_first does not retain dtypes with Timestamp DataFrames

### LINKED ISSUE #28481: combine_first returns unexpected results for timestamp dataframes
#### Problem Description
`combine_first` returns weird results in case there exist unmatched columns between two timestamp dataframes.  this issue probably relates to #24357 (not retaining dtypes)

```python
import pandas as pd
x = pd.DataFrame([pd.NaT,pd.NaT],columns=['a']).T
y = pd.DataFrame([[pd.NaT,pd.NaT],pd.to_datetime(['20190101','20190102'])],index=['a','b'],columns=[1,2])
```

<details>

x|0|1
---|---|---
a|NaT|NaT

y|1|2
---|---|---
a|NaT|NaT
b|2019-01-01|2019-01-02

</details>

##### Output of ``x.combine_first(y)``

current|0|1|2
---|---|---|---
a|NaT|NaT|-9.223372e+18
b|NaT|2019-01-01|1.546387e+18

##### Output of ``x.reindex(columns=x.columns|y.columns).combine_first(y)``
expected|0|1|2
---|---|---|---
a|NaT|NaT|NaT
b|NaT|2019-01-01|2019-01-02

#### INSTALLED VERSIONS
commit           : None
python           : 3.7.4.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.18.0-80.7.2.el8_0.x86_64
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.25.1
numpy            : 1.17.2

### PR DESCRIPTION
**Pull request by arw2019**
- [x] closes #28481
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Picking up #35514

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Why does combin_first use expressions.where instead of mgr.where? Casting should be handled correctly there

**Comment by arw2019** [comment]
> Why does combin_first use expressions.where instead of mgr.where? Casting should be handled correctly there

That line hasn't been touched since #17744, will look

**Comment by jbrockmendel** [comment]
Now that i have an actual keyboard: this would likely also require fixing Block.where #38073

**Comment by arw2019** [comment]
> Now that i have an actual keyboard: this would likely also require fixing Block.where #38073 

Ok - in this case this would sit on top of that

I'm struggling a little with the pattern. Atm I have:
``` python
    def combine_first(self, other: DataFrame) -> DataFrame:
        def combiner(x, y):
            mask = extract_array(isna(x))

            x_values = extract_array(x, extract_numpy=True)
            y_values = extract_array(y, extract_numpy=True)

            # If the column y in other DataFrame is not in first DataFrame,
            # just return y_values.
            if y.name not in self.columns:
                return y_values

            _where = self._mgr.where(y_values, mask, align=True, errors="raise", try_cast=True, axis=1)
            # extract array from _where

        return self.combine(other, combiner, overwrite=False)
```
Is this in the right ballpark?

File: pandas/tests/frame/methods/test_combine_first.py [hunk 1]
```diff
@@ -367,6 +392,25 @@ def test_combine_first_string_dtype_only_na(self):
         tm.assert_frame_equal(result, expected)
 
 
+@pytest.mark.parametrize(
+    "scalar1, scalar2",
+    [
+        (datetime(2020, 1, 1), datetime(2020, 1, 2)),
+        (pd.Period("2020-01-01", "D"), pd.Period("2020-01-02", "D")),
```
Comment (jreback) [inline pandas/tests/frame/methods/test_combine_first.py:399]: can add an Interval example here as well

**Review by jreback — CHANGES_REQUESTED** [review]
> Now that i have an actual keyboard: this would likely also require fixing Block.where #38073

@jbrockmendel is the current patch not good for 1.2? (agreed could use some refactoring).

**Comment by jbrockmendel** [comment]
> is the current patch not good for 1.2? (agreed could use some refactoring).

This looks good for now

**Comment by jreback** [comment]
@arw2019 ok small test comment. ping on green.

File: pandas/tests/frame/methods/test_combine_first.py (hunk 1 above)
Comment (arw2019) [reply pandas/tests/frame/methods/test_combine_first.py:399]: Done

**Comment by jreback** [comment]
thanks @arw2019

**Comment by jbrockmendel** [comment]
@arw2019 do you have time to try to make this go through `self._mgr.where`?  (no hurry)

**Comment by arw2019** [comment]
> @arw2019 do you have time to try to make this go through `self._mgr.where`? (no hurry)

yes, in the works

**Pull request merged**